xarray/core/groupby.py

```diff
@@ -294,92 +307,315 @@ def _apply_loffset(
+                closed=grouper.closed,
+                label=grouper.label,
+                origin=grouper.origin,
+                offset=grouper.offset,
+                loffset=grouper.loffset,
+            )
+        else:
+            index_grouper = pd.Grouper(
+                freq=grouper.freq,
+                closed=grouper.closed,
+                label=grouper.label,
+                origin=grouper.origin,
+                offset=grouper.offset,
+            )
+        self.index_grouper: CFTimeGrouper | pd.Grouper = index_grouper
+
+    def _get_index_and_items(self) -> tuple[pd.Index, pd.Series, np.ndarray]:
+        first_items, codes = self.first_items()
+        full_index = first_items.index
+        if first_items.isnull().any():
+            first_items = first_items.dropna()
+
+        full_index = full_index.rename("__resample_dim__")
+        return full_index, first_items, codes
+
+    def first_items(self) -> tuple[pd.Series, np.ndarray]:
+        from xarray import CFTimeIndex
+
+        if isinstance(self.group_as_index, CFTimeIndex):
+            return self.index_grouper.first_items(self.group_as_index)
+        else:
+            s = pd.Series(np.arange(self.group_as_index.size), self.group_as_index)
+            grouped = s.groupby(self.index_grouper)
+            first_items = grouped.first()
+            counts = grouped.count()
+            # This way we generate codes for the final output index: full_index.
+            # So for _flox_reduce we avoid one reindex and copy by avoiding
+            # _maybe_restore_empty_groups
+            codes = np.repeat(np.arange(len(first_items)), counts)
+            if self.grouper.loffset is not None:
+                _apply_loffset(self.grouper.loffset, first_items)
+            return first_items, codes
+
+    def factorize(self, squeeze: bool) -> None:
+        self.full_index, first_items, codes = self._get_index_and_items()
+        sbins = first_items.values.astype(np.int64)
+        self.group_indices = [slice(i, j) for i, j in zip(sbins[:-1], sbins[1:])]
+        self.group_indices += [slice(sbins[-1], None)]
+
+        self.unique_coord = IndexVariable(
+            self.group.name, first_items.index, self.group.attrs
         )
-    unique_coord = IndexVariable(group.name, unique_values)
-    return unique_coord, group_indices, codes
-
-
-def _factorize_dummy(
-    group, squeeze: bool
-) -> tuple[IndexVariable, T_GroupIndices, np.ndarray]:
-    # no need to factorize
-    group_indices: T_GroupIndices
-    if not squeeze:
-        # use slices to do views instead of fancy indexing
-        # equivalent to: group_indices = group_indices.reshape(-1, 1)
-        group_indices = [slice(i, i + 1) for i in range(group.size)]
+        self.codes = self.group.copy(data=codes)
+
+
+class Grouper(ABC):
+    pass
+
+
+@dataclass
+class UniqueGrouper(Grouper):
+    pass
+
+
+@dataclass
+class BinGrouper(Grouper):
+    bins: Any  # TODO: What is the typing?
```

Comment:
Never really like to use 'Sequence' because 'np.ndarray's are not Sequences... And I think it could be quite common to supply the bins with bumpy arrays?

Reply:
np.typing.ArrayLike?

Reply:
Yes but "nested  sequences" won't  work here:https://numpy.org/doc/stable/reference/typing.html#numpy.typing.ArrayLike